web_src/js/components/RepoActionView.vue

```diff
@@ -250,21 +285,38 @@ const sfc = {
       lineNumber.textContent = line.index;
       div.append(lineNumber);
 
-      // TODO: Support displaying time optionally
-
-      const logMessage = document.createElement('div');
+      // for "Show timestamps"
+      const logTimeStamp = document.createElement('relative-time');
+      logTimeStamp.className = 'log-time-stamp';
+      const date = new Date(parseFloat(line.timestamp * 1000));
+      logTimeStamp.setAttribute('datetime', date);
+      logTimeStamp.setAttribute('format', 'datetime');
+      logTimeStamp.setAttribute('time-zone-name', 'short');
+      toggleElem(logTimeStamp, this.timeVisible.stamp);
+      // for "Show log duration"
+      const logTimeDuration = document.createElement('span');
+      logTimeDuration.className = 'log-time-duration';
+      const duration = Math.floor(parseFloat(line.timestamp) - parseFloat(startTime));
+      logTimeDuration.textContent = `${duration}s`;
+      toggleElem(logTimeDuration, this.timeVisible.duration);
+
+      const logMessage = document.createElement('span');
       logMessage.className = 'log-msg';
       logMessage.innerHTML = ansiLogToHTML(line.message);
+      div.append(logTimeStamp);
       div.append(logMessage);
+      div.append(logTimeDuration);
 
       return div;
     },
 
     appendLogs(stepIndex, logLines) {
+      if (!logLines.length) return;
+      const startTime = logLines[0].timestamp;
       for (const line of logLines) {
         // TODO: group support: ##[group]GroupTitle , ##[endgroup]
         const el = this.getLogsContainer(stepIndex);
-        el.append(this.createLogLine(line));
+        el.append(this.createLogLine(line, startTime));
```

Comment:
Is it right to do so? Why `logLines[0]` is the `startTime`.

I am not sure whether backend would send partial logs by "tail".

Reply:
I am not very sure, but If the order of log lines is sent by backend correctly, then the timestamp of the first line of the step should be the starting timestamp of a log step.

Reply:
Suppose:

* log-a  01:02:03
* log-b  01:02:04
* log-c  01:02:09

What's `log-c` 's duration? 6 or 5 ?

Reply:
It's 6, the duration is current log timestamp compared to the starting timestamp of the current log step. with reference to [dronw ci](https://drone.gitea.io/go-gitea/gitea/74864/1/5).

Reply:
I guess this line is the `logLines[0]`  in drone's logs.

That's what I asked about: whether backend would send partial logs by "tail" ?

Assuming `logLines[0]` is the starting line seems fragile. Backend should tell frontend the accurate start time.


![image](https://github.com/go-gitea/gitea/assets/2114189/f10c6c23-dcde-4ff3-8cf6-9076b0492c42)

Reply:
And the name "duration" seems not the right concept .....  it is not the log's duration, it is the log's time offset (or some other name) to the start time.

Definition: duration: the time during which something continues. The single log itself doesn't continue 31s, but it's time is 31s after the start time.

Reply:
> I guess this line is the `logLines[0]` in drone's logs.
> 
> That's what I asked about: whether backend would send partial logs by "tail" ?
> 
> Assuming `logLines[0]` is the starting line seems fragile. Backend should tell frontend the accurate start time.
> 
> ![image](https://user-images.githubusercontent.com/2114189/241111778-f10c6c23-dcde-4ff3-8cf6-9076b0492c42.png)

I think it is fragile when the job is still running, as the logs keep appending to the logs. But I am not very sure where to add the start time for the certain log step at backend, I think it should be somewhere around [here](https://github.com/go-gitea/gitea/blob/e4922d484b9ee94ba22e5ff08b3c25c8be09d9c8/routers/web/repo/actions/view.go#L212). Might need some help with this..

Reply:
Done in https://github.com/go-gitea/gitea/pull/24876/commits/15dbbd8b7f17700baf6cd57c2ee4eab55381d29e